// @comunica/actor-dereference-file/lib/ActorDereferenceFile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActorDereferenceFile = void 0;
const fs_1 = require("fs");
const url_1 = require("url");
const bus_dereference_1 = require("@comunica/bus-dereference");
/**
 * A comunica File Dereference Actor.
 */
class ActorDereferenceFile extends bus_dereference_1.ActorDereference {
    constructor(args) {
        super(args);
    }
    async test({ url }) {
        try {
            (0, fs_1.accessSync)(getPath(url), fs_1.constants.F_OK);
        }
        catch (error) {
            throw new Error(`This actor only works on existing local files. (${error})`);
        }
        return true;
    }
    static isURI(str) {
        const URIRegex = /\w[\w+.-]*:.*/u;
        return URIRegex.exec(str) !== null;
    }
    async run({ url }) {
        const requestTimeStart = Date.now();
        return {
            data: (0, fs_1.createReadStream)(getPath(url)),
            // This should always be after the creation of the read stream
            requestTime: Date.now() - requestTimeStart,
            exists: true,
            url: ActorDereferenceFile.isURI(url) ? url : (0, url_1.pathToFileURL)(url).href,
        };
    }
}
exports.ActorDereferenceFile = ActorDereferenceFile;
const getPath = (url) => url.startsWith('file://') ? (0, url_1.fileURLToPath)(url) : url;
